pkg/chunk/aws_storage_client.go

```diff
@@ -331,28 +340,232 @@ func (a dynamoDBRequestAdapter) HasNextPage() bool {
 	return a.request.HasNextPage()
 }
 
-func (a awsStorageClient) GetChunk(ctx context.Context, key string) ([]byte, error) {
+func (a awsStorageClient) GetChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	var (
+		s3Chunks       []Chunk
+		dynamoDBChunks []Chunk
+	)
+
+	for _, chunk := range chunks {
+		if chunk.From.Before(a.cfg.ChunkTableFrom.Time) {
+			s3Chunks = append(s3Chunks, chunk)
+		} else {
+			dynamoDBChunks = append(dynamoDBChunks, chunk)
+		}
+	}
+
+	var err error
+	s3Chunks, err = a.getS3Chunks(ctx, s3Chunks)
+	if err != nil {
+		return nil, err
+	}
+
+	dynamoDBChunks, err = a.getDynamoDBChunks(ctx, dynamoDBChunks)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(dynamoDBChunks, s3Chunks...), nil
+}
+
+func (a awsStorageClient) getS3Chunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	incomingChunks := make(chan Chunk)
+	incomingErrors := make(chan error)
+	for _, chunk := range chunks {
+		go func(chunk Chunk) {
+			chunk, err := a.getS3Chunk(ctx, chunk)
+			if err != nil {
+				incomingErrors <- err
+				return
+			}
+			incomingChunks <- chunk
+		}(chunk)
+	}
+
+	result := []Chunk{}
+	errors := []error{}
+	for i := 0; i < len(chunks); i++ {
+		select {
+		case chunk := <-incomingChunks:
+			result = append(result, chunk)
+		case err := <-incomingErrors:
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return nil, errors[0]
+	}
+	return result, nil
+}
+
+func (a awsStorageClient) getS3Chunk(ctx context.Context, chunk Chunk) (Chunk, error) {
 	var resp *s3.GetObjectOutput
 	err := instrument.TimeRequestHistogram(ctx, "S3.GetObject", s3RequestDuration, func(ctx context.Context) error {
 		var err error
 		resp, err = a.S3.GetObjectWithContext(ctx, &s3.GetObjectInput{
 			Bucket: aws.String(a.bucketName),
-			Key:    aws.String(key),
+			Key:    aws.String(chunk.externalKey()),
 		})
 		return err
 	})
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
 	}
 	defer resp.Body.Close()
 	buf, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, err
+		return Chunk{}, err
+	}
+	if err := chunk.decode(buf); err != nil {
+		return Chunk{}, err
 	}
-	return buf, nil
+	return chunk, nil
 }
 
-func (a awsStorageClient) PutChunk(ctx context.Context, key string, buf []byte) error {
+func (a awsStorageClient) getDynamoDBChunks(ctx context.Context, chunks []Chunk) ([]Chunk, error) {
+	outstanding := dynamoDBReadRequest{}
+	chunksByKey := map[string]Chunk{}
+	for _, chunk := range chunks {
+		key := chunk.externalKey()
+		chunksByKey[key] = chunk
+		tableName := a.chunkTableFor(chunk.From)
+		outstanding.Add(tableName, key, nil)
+	}
+
+	result := []Chunk{}
+	unprocessed := dynamoDBReadRequest{}
+	backoff, numRetries := minBackoff, 0
+	for outstanding.Len()+unprocessed.Len() > 0 {
+		requests := dynamoDBReadRequest{}
+		requests.takeReqs(unprocessed, dynamoDBMaxReadBatchSize)
+		requests.takeReqs(outstanding, dynamoDBMaxReadBatchSize)
```

Comment:
>  It's not used in the actual request to DynamoDB (which uses outstanding instead)

That was a mistake - it should have used requests.

The idea is that we have two sets of requests - the original ones we need to do, and the ones we need to retry.  Every time round the loop we take the ones we need to retry first (to give them priority, to prevent a long tail), and then fill the request up with original ones.  When the request has a partial failure, we take what was left and put it back on the unprocessed list.

This is a mirror of the way we do writes, and is complicated.  But this is DynamoDB.